src/python/pants/engine/legacy/graph.py

```diff
@@ -269,6 +275,93 @@ def _inject_specs(self, subjects):
       yield hydrated_target.address
 
 
+class _DependentGraph(object):
+  """A graph for walking dependent addresses of TargetAdaptor objects.
+
+  This avoids/imitates constructing a v1 BuildGraph object, because that codepath results
+  in many references held in mutable global state (ie, memory leaks).
+
+  The long term goal is to deprecate the `changed` goal in favor of sufficiently good cache
+  hit rates, such that rather than running:
+
+    ./pants --changed-parent=master test
+
+  ...you would always be able to run:
+
+    ./pants test ::
+
+  ...and have it complete in a similar amount of time by hitting relevant caches.
+  """
+
+  @classmethod
+  def from_iterable(cls, target_types, address_mapper, adaptor_iter):
+    """Create a new DependentGraph from an iterable of TargetAdaptor subclasses."""
+    inst = cls(target_types, address_mapper)
+    all_valid_addresses = set()
+    for target_adaptor in adaptor_iter:
+      inst._inject_target(target_adaptor)
+      all_valid_addresses.add(target_adaptor.address)
+    inst._validate(all_valid_addresses)
+    return inst
+
+  def __init__(self, target_types, address_mapper):
+    # TODO: Dependencies and implicit dependencies are mapped independently, because the latter
+    # cannot be validated until:
+    #  1) Subsystems are computed in engine: #5869. Currently instantiating a subsystem to find
+    #     its injectable specs would require options parsing.
+    #  2) Targets-class Subsystem deps can be expanded in-engine (similar to Fields): #4535,
+    self._dependent_address_map = defaultdict(set)
+    self._implicit_dependent_address_map = defaultdict(set)
+    self._target_types = target_types
+    self._address_mapper = address_mapper
+
+  def _validate(self, all_valid_addresses):
+    """Validate that all of the dependencies in the graph exist in the given addresses set."""
+    for dependency, dependents in iteritems(self._dependent_address_map):
+      if dependency not in all_valid_addresses:
+        raise AddressLookupError(
+            'Dependent graph construction failed: {} did not exist. Was depended on by:\n  {}'.format(
+             dependency.spec,
+             '\n  '.join(d.spec for d in dependents)
+          )
+        )
+
+  def _inject_target(self, target_adaptor):
```

Comment:
It would be nice to hoist `_inject_target` and `_validate` into classmethods / staticmethods that prepared the dicts to keep the factory / instance line sharp.

Reply:
Agreed, but this was the previous impl... going to leave for now.